src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -393,37 +366,162 @@ public async Task<uint256> Rewind()
+                    }
 
-                    this.logger.LogTrace("(-)[REWOUND_TO_INNER]:'{0}'", this.blockHash);
-                    return this.blockHash;
-                }
+                    this.rewindDataBatch.Remove(lastItem);
+
+                    this.logger.LogTrace("(-)[REMOVED_FROM_BATCH]:'{0}'", lastItem.RewindData.PreviousBlockHash);
+                    return lastItem.RewindData.PreviousBlockHash;
+                } 
 
-                uint256 hash = await this.inner.Rewind().ConfigureAwait(false);
-                this.innerBlockHash = hash;
+                uint256 hash = await this.CoinViewStorage.Rewind().ConfigureAwait(false);
+                this.persistedBlockHash = hash;
                 this.blockHash = hash;
 
                 this.logger.LogTrace("(-):'{0}'", hash);
                 return hash;
             }
         }
 
+        /// <summary>
+        /// Dequeues the rewind data continuously and saves it to the database when max batch size is reached or timer ran out.
+        /// </summary>
+        /// <remarks>Batch is always saved on shutdown.</remarks>
+        private async Task DequeueRewindDataContinuouslyAsync()
+        {
+            this.logger.LogTrace("()");
+
+            Task<QueuedRewindData> dequeueTask = null;
+            Task timerTask = null;
+
+            while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+            {
+                // Start new dequeue task if not started already.
+                dequeueTask = dequeueTask ?? this.rewindDataQueue.DequeueAsync();
+
+                // Wait for one of the tasks: dequeue or timer (if available) to finish.
+                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new rewind data 
+                // or the max batch size is reached or the node is shutting down.
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    QueuedRewindData item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    this.rewindDataBatch.Add(item);
+
+                    this.currentCountOfAllRewindDataItems += item.ItemsCount;
```

Comment:
In a recent PR I did @Aprogiena asked me to use `Interlocked.Add` when keeping count of an `AsyncQueue`.

https://github.com/stratisproject/StratisBitcoinFullNode/blob/39e7a38e70ef9cf1ee4b1cde1b0c7843736cf9a1/src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs#L268-L273

I'm not 100% sure you would need it here - @Aprogiena ?

Also we added trace logging around any change in queue size - which was useful.  Maybe you can add some around here as well.